Andropov_Vsevolod

heptaber opened the pull request.

heptaber committed:
upd

heptaber committed:
upd ReadMe.md

NikolaevArtem commented on README.md:
```diff
@@ -1,9 +1,7 @@
-| HW1 | [Console printer](https://github.com/NikolaevArtem/Java_Core_June_2021/tree/master/src/main/java/homework_1) | The app that reads input arguments and prints them, until "error" argument |
-
-[Link to markdown giude](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)
+| HW1 | [Console printer](https://github.com/NikolaevArtem/Java_Core_June_2021/tree/master/src/main/java/homework_1) | [hw1_solution](./src/main/java/homework_1/) |
```
It still looks at my PR

heptaber replied:
changed.

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -3,7 +3,27 @@
-    System.out.println("Hello homework!");
+    for (String arg : args) {
+      if (arg.equals("ошибка") || arg.equals("error")) {
+        System.err.println("Тревога!");
```
Doesn't work as expected
![image](https://user-images.githubusercontent.com/58872509/125207756-af5cd300-e296-11eb-99c1-edfedcdd4c0f.png)

heptaber committed:
readme fix.

heptaber committed:
upd

heptaber committed:
upd

heptaber committed:
upd

heptaber committed:
upd

heptaber committed:
homework_2

heptaber committed:
PyramidPrinter optimised:
+ Using 2 functions that makes able to faster processing;
+ Using Math.abs for input value for being sure input number is positive (previous version was just stopping program execution by assert);
+ Assert now just checking input value is not zero;
- Current version using much more memory then previous;

heptaber committed:
codingbat profile added

heptaber committed:
_

heptaber committed:
Merge remote-tracking branch 'origin/master' into feature/AndropovVsevolod

heptaber committed:
hw3 partly done

heptaber committed:
builder.gradle lombok added
immutable class basic created

heptaber committed:
changed tests directories; fixed gradle testing.

heptaber committed:
refactored a little bit

heptaber committed:
upd

heptaber committed:
HW3 added, changed paths to HW2's.

heptaber committed:
parametrized test implemented

heptaber committed:
valid tests added

heptaber committed:
structure changed

heptaber committed:
101

heptaber committed:
deleted extra import ^_^

heptaber committed:
small fixes

heptaber committed:
_

heptaber committed:
_

heptaber committed:
refactored: added parametrized test for not valid data;

heptaber committed:
_

heptaber committed:
refactored tests to parametrized

NikolaevArtem commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+
+import static homework_2.IOMod.*;
+
+public class RandomCharsTable {
```
![image](https://user-images.githubusercontent.com/58872509/126909996-3a5bbcb3-0d85-445b-9158-160a12e12fd7.png)

NikolaevArtem commented:
Good! The apps work "almost" as expected, I'll continue review when I have time

heptaber committed:
text file add

heptaber committed:
hw 4

heptaber committed:
small fix

heptaber committed:
add annotation test

pepya4ka commented on README.md:
```diff
@@ -1,9 +1,20 @@
+HW3   | [Immutable Class](./src/main/java/homework_3/ImmutableClass.java) | Create Immutable class: several fields, including reference type; several constructors; method returning a new object. | 
+HW4.1 | [Custom File Reader](./src/main/java/homework_4/custom_file_reader/) <br/> [Tests](./src/test/java/homework_4/custom_file_reader/) | App reading data from file and prints to console. Text file path: main/resources/custom_file_reader/. Reading file in 3 ways (one of them using NIO) executable with methods: run1(), run2(), run3(). |
+HW4.2 | [Singleton](./src/main/java/homework_4/singleton/) <br/> [Tests](./src/test/java/homework_4/singleton/) | Singleton App: private constructor and public method getInstance() |
+HW4.3 | [Custom Annotation](./src/main/java/homework_4/custom_annotation/) <br/> [Tests](./src/test/java/homework_4/custom_annotation/) | Custom annotation class. |
```
tests - "page not found"

heptaber replied:
not added annotation test yet. will be active soon :)

pepya4ka commented on src/main/java/homework_1/Main.java:
```diff
@@ -2,8 +2,18 @@
-  public static void main(String[] args) {
-    System.out.println("Hello homework!");
-  }
+    static final String RED = "\u001b[31m";
```
fix access modifier

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+        String s = bufferedReaderStringReader();
+
+        try {
+            String[] parameters = s.split(" ");
```
Попробуй создать метод `validInput`, например, который будет выбрасывать исключение `IllegalArgumentException` если какое-либо из условий не выполняется. По-моему, это поможет избавиться от валидации и от вызовов sout внутри try.

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+            } else {
+                System.out.println(FORMAT_ERROR);
+                return;
+            }
```
```suggestion
            boolean strategy = !parameters[2].equalsIgnoreCase("odd");
```
Если будет метод `validInput`, конструкцию можно заменить одной строчкой

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+        if ((ch % 2 == 1) && (!String.valueOf(result).contains(String.valueOf(ch)))) {
+            result.append(ch);
+        }
+    }
```
```suggestion
        private void generateRes(int strategy, char ch, StringBuilder result) {
        if ((ch % 2 == strategy) && (!String.valueOf(result).contains(String.valueOf(ch)))) {
            result.append(ch);
        }
    }
```
Можно заменить одним методом, и вместо конструкции ifelse использовать одну строчку 
`generateRes(strategy ? 0 : 1, randAbc[i][j], result);`

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    private void printResult(boolean strategy, StringBuilder result) {
```
Если я не ошибаюсь, на одной из лекции оговаривали, что если в таблице нет подходящих символов, то все равно выводим "Even\odd letters - "

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+            System.out.print("Odd letters - ");
+        }
+        int i = 0;
+        while (i < result.length()) { // to avoid exception in a case when result.length() = 0 ;)
```
Почему не цикл for?

pepya4ka commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,106 @@
+        }
+        int i = 0;
+        while (i < result.length()) { // to avoid exception in a case when result.length() = 0 ;)
+            if (i == result.length() - 1) {
+                System.out.print(result.charAt(i));
+                break;
+            }
+            System.out.print(result.charAt(i) + ", ");
```
```suggestion
            System.out.print(result.charAt(i));
            if (i != result.length() - 1) {
                System.out.print(", ");
            }
```
Так будет работать?

pepya4ka commented on src/main/java/homework_2/traffic_light/TrafficLight.java:
```diff
@@ -0,0 +1,47 @@
+import static homework_2.IOMod.*;
+import static homework_2.Colors.*;
+
+public class TrafficLight {
```
![image](https://user-images.githubusercontent.com/44002713/128746802-ddce392d-bc44-4bad-9aa5-28e7ca82078b.png)

heptaber replied:
fixed

pepya4ka commented on src/main/java/homework_2/traffic_light/ExtraTrafficLight.java:
```diff
@@ -0,0 +1,42 @@
+
+    private void colorPick(int secs) {
+        // 0 <= light < 35 green, 35 <= light < 40 yellow, 40 <= light < 55 red, 55 <= light < 60 yellow
+        if (secs >= 0 && secs < 35) {
```
![image](https://user-images.githubusercontent.com/44002713/128747370-4e57a0d8-4f14-4938-a382-87322c86df1a.png)

pepya4ka commented on src/main/java/homework_2/traffic_light/TrafficLight.java:
```diff
@@ -0,0 +1,47 @@
+            int secondsGone = IOMod.bufferedReaderIntReader();
+            if (secondsGone < 0) {
+                System.out.println(ERROR);
+                return;
```
return не нужен здесь

pepya4ka commented on src/main/java/homework_2/traffic_light/TrafficLight.java:
```diff
@@ -0,0 +1,47 @@
+        inputSeconds();
+    }
+
+    private void inputSeconds() {
```
Можно сделать аналогично с `RandomCharsTable`

pepya4ka commented on src/test/java/homework_2/pyramid_printer/sources/PyramidPrintTest.java:
```diff
@@ -0,0 +1,74 @@
+    PyramidPrinter pyramidPrinter = new PyramidPrinter();
+
+    // testing valid input
+    private static Stream<Arguments> validCases() {
```
Опустить пониже в классе

pepya4ka commented on src/test/java/homework_2/pyramid_printer/sources/PyramidPrintTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    // testing not valid input values
+    private static Stream<Arguments> notValidCases() {
```
Опустить пониже в классе

heptaber committed:
upd

heptaber committed:
upd

heptaber committed:
global hw4 refactoring;

heptaber committed:
upd

heptaber committed:
mockito dependency added

heptaber committed:
empty file: processing removed, test removed;
text file renamed;

heptaber committed:
recursion power added;
test power added;

heptaber committed:
negative power test added;
negative power ability has been implemented;

heptaber committed:
armstrong number realization;
https://en.wikipedia.org/wiki/Narcissistic_number

heptaber committed:
prints method name fix;

heptaber committed:
armstrong number computation app removed;

heptaber committed:
org.jetbrains:annotations updated to 22.0.0;
mockito removed;

heptaber committed:
custom regex matcher added;
tests for custom regex matcher added;

heptaber committed:
PowerOfNumber small fixes;

heptaber committed:
hw5 added;

heptaber committed:
hw6 added;

heptaber committed:
hw6 tests added;

heptaber committed:
interface renamed;
interface method removed;

heptaber committed:
hw6, hw6 tests added;

heptaber committed:
hw6 upd;
hw6 tests removed;
readme upd;

heptaber committed:
methods name refactored;
hw6 refactored, structure changed;

heptaber committed:
hw7 added

heptaber committed:
hw7 added

heptaber committed:
hw7 refactored;
hw7 tests added;

heptaber committed:
hw7 upd;

heptaber committed:
little folders reshuffle ;)
readme upd

heptaber committed:
little fix

heptaber committed:
course project: SeaBattle add;

heptaber committed:
changed field size to 10 ;)
changed ship naming functions;
README upd;

heptaber committed:
upd for more comfortable PVP game

heptaber committed:
bug fixed :)

NikolaevArtem commented on src/main/java/course_project/Main.java:
```diff
@@ -0,0 +1,11 @@
+
+public class Main {
+
+    public static void main(String[] args) {
```
It seems indentation is shifted
![image](https://user-images.githubusercontent.com/58872509/134234647-69ab186e-ffd0-4818-84fc-3def141659eb.png)

heptaber replied:
can you describe this case in detail?
my view:
<img width="722" alt="Screen Shot 2021-09-21 at 10 28 39 PM" src="https://user-images.githubusercontent.com/26281676/134235335-81ef33d2-923a-4443-83f1-4451893c2608.png">

NikolaevArtem replied:
Well done! The game interface looks pretty nice, even though it lacks row delimiters. Playing is comfortable, it shows pretty informative messages. Also, a lot of features are introduced, as well as exit code and description. You show good knowledge of Java Core and Java libraries, abstractions are mostly well-picked, architecture is good, but it can be improved. Also, you have lots of static, and it's not recommended. 

Approved!

heptaber replied:
I understood you. Thanks for the review!

NikolaevArtem commented on src/main/java/course_project/services/SeaBattle.java:
```diff
@@ -0,0 +1,26 @@
+
+    public void run() {
+        DialogueMenu.printGreetings();
+        DialogueMenu.printMainMenu();
```
The method name doesn't correlate with what it does

heptaber replied:
Fully agree, but i couldn't find better for this.
Before the game process started it's fully controlled by the menu - so User picks whatever wants,  that's why it's like an entrance point into the game at this moment.

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,51 @@
+
+public final class ImmutableClass {
+
+    private final String name; // reference type
```
I know it's unclear in description, but I meant field which it mutable...

NikolaevArtem commented on src/main/java/homework_4/custom_file_reader/CustomFileReader.java:
```diff
@@ -0,0 +1,67 @@
+import static homework_4.custom_file_reader.utils.ModifiedStringPrinter.printModifiedString;
+
+
+public class CustomFileReader {
```
Good abstraction!

NikolaevArtem commented on src/main/java/homework_4/singleton/Main.java:
```diff
@@ -0,0 +1,15 @@
+ */
+
+
+public class Main {
```
Quite nice approach to HW_4! Looks good

NikolaevArtem commented on src/main/java/homework_5/custom_regex_matcher/CustomRegexMatcher.java:
```diff
@@ -0,0 +1,27 @@
+
+    private CustomRegexMatcher() { }
+
+    private static final String PHONE_NUMBER_FORMAT = "(\\+\\d)\\s\\(\\d{3}\\)\\s(\\d{3})-(\\d{4})";
```
Opt: You could add some variability to regex (like, accept phone numbers with different country codes and etc)

NikolaevArtem commented on src/main/java/homework_6/map_problems_collision_generator/MapProblemsMutableGenerator.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public int hashCode() {
+        return new Random().nextInt(Integer.MAX_VALUE - 1);
```
Well, there;s no mutable key problem, it's just invalid hashcode implementation. Mutable key generates problem even if hashcode impl is valid

blowushka replied:
He fixed hashCode() and seems like mutable problem is appeared!

NikolaevArtem commented on src/main/java/homework_7/kitten_to_cat/Cat.java:
```diff
@@ -0,0 +1,56 @@
+
+public class Cat {
+
+    private final int age;
+    private final String name;
```
Opt: you could provide such fields and such lambda, that it'll make type conversion

NikolaevArtem reviewed (changes requested):
Well-done! You introduce new approaches to tasks, experiment with Java Core, which is cool. There are minor issues to fix, but in general, your HW looks really good! And special thanks for all code review efforts you did

heptaber committed:
homeworks fixes

heptaber committed:
course project optimizations

heptaber committed:
hw3 upd

heptaber committed:
hw3 upd

blowushka commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -48,7 +49,7 @@ private void initAllDataList() {
         return new LinkedList<>(this.allDataList);
```
Good approach! Approved!
